# V. Harvey IBM Data Science Specialization Capstone Project Notebook
## Peer-Graded Assignment: Segmenting and Clustering Neighborhoods in Toronto
### Part 1: Questions 1 - 4: Create DataFrame

#### Import Pandas Library:

In [1]:
import pandas as pd

#### Read HTML table:
Get first table

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df = pd.read_html(url)
df_toronto = df[0]
df_toronto

Post Code           Borough  \
0         M1A      Not assigned   
1         M2A      Not assigned   
2         M3A        North York   
3         M4A        North York   
4         M5A  Downtown Toronto   
..        ...               ...   
175       M5Z      Not assigned   
176       M6Z      Not assigned   
177       M7Z      Not assigned   
178       M8Z         Etobicoke   
179       M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

### Exploring the dataset:

In [3]:
df_toronto.shape

(180, 3)

In [4]:
df_toronto.columns

Index(['Post Code', 'Borough', 'Neighborhood'], dtype='object')

**Note: the uncleaned dataframe has 180 rows and 3 columns**

**Getting column metadata:**

In [5]:
df_toronto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Post Code     180 non-null    object
 1   Borough       180 non-null    object
 2   Neighborhood  180 non-null    object
dtypes: object(3)
memory usage: 2.2+ KB


**Each row has 180 non-null values, corresponding to the number of rows in the dataframe. All having the 'object' datatype, which is what we would expect.**

In [6]:
df_toronto.head()

Post Code           Borough               Neighborhood
0       M1A      Not assigned               Not assigned
1       M2A      Not assigned               Not assigned
2       M3A        North York                  Parkwoods
3       M4A        North York           Victoria Village
4       M5A  Downtown Toronto  Regent Park, Harbourfront

**However, unassigned values are indicated with the value "Not assigned". The dataframe includes values for column "Borough" that are "Not assigned." These must be removed from the dataframe.**

### Clear all "Not assigned" Boroughs

In [7]:
df_toronto['Borough'].value_counts()

Not assigned        77
North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
York                 5
East York            5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

**There are 77 "Not assigned" Boroughs**

#### Get indices for rows where Borough has value "Not assigned":

In [8]:
df_toronto_indexNames = df_toronto[df_toronto['Borough'] == "Not assigned"].index
df_toronto_indexNames

Int64Index([  0,   1,   7,  10,  15,  16,  19,  24,  25,  28,  29,  33,  34,
             35,  37,  38,  42,  43,  44,  51,  52,  53,  60,  61,  62,  69,
             70,  71,  78,  79,  87,  88,  96,  97, 101, 105, 106, 110, 115,
            118, 119, 123, 124, 125, 127, 128, 131, 132, 133, 134, 136, 137,
            140, 141, 145, 146, 149, 150, 154, 155, 158, 159, 161, 162, 163,
            164, 166, 167, 170, 171, 172, 173, 174, 175, 176, 177, 179],
           dtype='int64')

#### Delete these row indices from dataFrame:

In [9]:
df_toronto_clean = df_toronto.drop(df_toronto_indexNames , inplace=False)

#### Display Dataframe:

In [10]:
df_toronto_clean

Post Code           Borough  \
2         M3A        North York   
3         M4A        North York   
4         M5A  Downtown Toronto   
5         M6A        North York   
6         M7A  Downtown Toronto   
..        ...               ...   
160       M8X         Etobicoke   
165       M4Y  Downtown Toronto   
168       M7Y      East Toronto   
169       M8Y         Etobicoke   
178       M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

**"Not assigned" rows have been deleted. There are now 103 rows (180 full rows - 77 "Not assigned" rows = 103 remaining)**

In [11]:
df_toronto_neighborhood = df_toronto_clean['Neighborhood'] == "Not assigned"
df_toronto_neighborhood

2      False
3      False
4      False
5      False
6      False
       ...  
160    False
165    False
168    False
169    False
178    False
Name: Neighborhood, Length: 103, dtype: bool

**Note: No neighborhood has a "Not assigned" value**

## FINAL Cleaned DataFrame:

**Reset indices:**

In [12]:
df_toronto_clean = df_toronto_clean.reset_index(drop=True)
df_toronto_clean

Post Code           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
..        ...               ...   
98        M8X         Etobicoke   
99        M4Y  Downtown Toronto   
100       M7Y      East Toronto   
101       M8Y         Etobicoke   
102       M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [13]:
df_toronto_clean.shape

(103, 3)